### PR TITLE
Integrate http tunnel

### LINKED ISSUE #36: Tunnel discussion
- look how  https://github.com/mmatczuk/go-http-tunnel and https://github.com/inconshreveable/ngrok works
### Current setup:
- run server on unsecure port
- accept daemon and user request on unsecure port

This is bad as we sent auth details over unsecure connection and we are open to mitm attacks

### Proposal #1
- http server
           - port chosen by user
           - we can support running on 443 with user provided certificate (on ice until we finish project)
           - this would expect user to provide reverse proxy for load balancer and certificate on reverse proxy
           - we would just serve rpc requests over this endpoint
- tcp tunnel server
           - port chosen by user
           - establishes long lasting tcp connection with daemon
           - all communications with daemons
Current issues:
- how to provide secure communication with daemons (encrypted tunnel connection)?
   - using go tls, we could postpone this later on and allow insecure for now
- refactoring of current daemon<->lb communications?
   - we could still send json over tcp but it would need envelope to determine msg type (see ngrok: https://github.com/inconshreveable/ngrok/blob/master/src/ngrok/msg/pack.go) 
   - or keep communication over http server
      - how do we authenticate peers on tunnel server?

### Proposal #2 
Use https://github.com/fatedier/frp

Issues:
- we are exposing user service publicly where anyone can access it directly

@mpetrun5 @MakMuftic Please look into this and let me know wdyt, I'm open to your proposals as well.

### LINKED ISSUE #51: Integrate go-http-tunel implementation
As a result of discussion #36, go-http-tunnel lib should be integrated inside the repo itself.

### PR DESCRIPTION
**Pull request by MakMuftic**
**Short description of work done**

Integrate and customize  [go-http-tunel](https://github.com/mmatczuk/go-http-tunnel) inside repo. Used [fork](https://github.com/datadope-io/go-http-tunnel/tree/added_port_auto_assignation) with autoassign port on server side.

### PR Checklist

- [X] I have run linter localy
- [X] I have run unit and integration tests locally
- [X] Rebased to master branch / merged master

### Changes

- Refactored auth to use token inside `X-Auth-Header`
- Restructured code and expose new (simpler) functions for creating `Client` and `Server`
- Refactor logging to be compatible with the rest of the repo
- Remove unnecessary code such as cmd commands from the original repo
- Refactor tests
- Tidy `go.mod` file

#### ~~Creating server~~
```go
s, err := server.NewServer(&server.ServerConfig{
	TlsCrtFilePath: "./pkg/http-tunel/testdata/selfsigned.crt",
	TlsKeyFilePath: "./pkg/http-tunel/testdata/selfsigned.key",
	Address:        ":5223",
	PortRange:      "10000:50000",
	AuthHandler: func(s string) bool {
		return s == "test-token"
	},
	Logger: log.NewEntry(l),
})

if err != nil {
	l.Fatal("failed to create server: %s", err)
}

s.Start()
```
#### ~~Creating client~~
```go
c, err := client.NewClient(&client.ClientConfig{
	ServerAddress:  "127.0.0.1:5223",
	TlsCrtFilePath: "./pkg/http-tunel/testdata/selfsigned.crt",
	TlsKeyFilePath: "./pkg/http-tunel/testdata/selfsigned.key",
	Tunnels: map[string]*client.Tunnel{
		"": {
			Protocol:   "tcp",
			Addr:       "localhost:3000",
			Auth:       "",
			Host:       "",
			RemoteAddr: "0.0.0.0:AUTO",
		},
	},
	Logger:    log.NewEntry(l),
	AuthToken: "test-token",
})

if err != nil {
	l.Fatal("failed to create client: %s", err)
}

if err := c.Start(); err != nil {
	l.Fatal("failed to start tunnels: %s", err)
}
```

### Issues

Closes #51 - issue
Closes #36 - original discussion

### CODE REVIEW

File: pkg/http-tunnel/client/client.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	clientData.proxy = CreateProxy(config.Tunnels, logger)
+
+	clientData.backoff = ExpBackoff(BackoffConfig{
+		Interval:    DefaultBackoffInterval,
+		Multiplier:  DefaultBackoffMultiplier,
+		MaxInterval: DefaultBackoffMaxInterval,
+		MaxTime:     DefaultBackoffMaxTime,
+	})
+
+	clientData.idName = "" // TODO
```
Comment (mpetrun5) [inline pkg/http-tunnel/client/client.go]: ?

File: pkg/http-tunnel/client/client.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+
+	return newClient(clientData)
+}
+
+func newClient(config *clientData) (*Client, error) {
+	if config.serverAddr == "" {
```
Comment (mpetrun5) [inline pkg/http-tunnel/client/client.go]: Does it make sense to check all of these fields below again when we checked them or assigned them in NewClient?

File: pkg/http-tunnel/client/client.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+	Tunnels map[string]*proto.Tunnel
+}
+
+func (c *Client) handleHandshake(w http.ResponseWriter, r *http.Request) {
+	c.logger.Infof("handshake for address %s", r.RemoteAddr)
+
+	w.Header().Add("X-Auth-Header", c.config.authToken)
+	w.WriteHeader(http.StatusOK)
+
+	// te := TunnelExt{
```
Comment (mpetrun5) [inline pkg/http-tunnel/client/client.go]: ?

File: pkg/http-tunnel/proto/controlmsg.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	if msg.Action == "" {
+		missing = append(missing, HeaderAction)
+	}
+	if msg.ForwardedHost == "" {
+		missing = append(missing, HeaderForwardedHost)
+	}
+	if msg.ForwardedProto == "" {
+		missing = append(missing, HeaderForwardedProto)
+	}
+	//if msg.ForwardedId == "" {
```
Comment (mpetrun5) [inline pkg/http-tunnel/proto/controlmsg.go:71]: Same as above.
If we don't need this logic might as well erase it.

File: pkg/http-tunnel/server/addrpool.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type RemoteID struct {
+	//ConnID   id.ID
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/addrpool.go]: ?

File: pkg/http-tunnel/server/pool.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+func newConnPool(t *http2.Transport, f func(identifier string)) *connPool {
+	return &connPool{
+		t:     t,
+		free:  f,
+		conns: make(map[string]connPair),
+	}
+}
+
+/*func (p *connPool) URL(identifier id.ID) string {
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/pool.go]: ?

File: pkg/http-tunnel/server/pool.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+}*/
+
+func (p *connPool) URL(identifier string) string {
+	return fmt.Sprint("https://", identifier)
+}
+
+func (p *connPool) GetClientConn(req *http.Request, addr string) (*http2.ClientConn, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	//fmt.Printf("GET CLIENNNNNNNNT CONNNNNNNN---------%s : ", addr)
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/pool.go]: Client goes BRRRRRRRRRRRRRRRRRR :joy:

File: pkg/http-tunnel/server/pool.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	if cp, ok := p.conns[addr]; ok {
+		p.close(cp, addr)
+	}
+}
+
+func (p *connPool) Ping(identifier string) (time.Duration, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	//addr := p.addr(identifier)
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/pool.go]: ?

File: pkg/http-tunnel/server/pool.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	return nil
+}
+
+func (p *connPool) DeleteConn(identifier string) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	//addr := p.addr(identifier)
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/pool.go]: ?

File: pkg/http-tunnel/server/registry.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.items[cname]; ok {
+		r.logger.Errorf("error on subscribe to registry this entry already exist %s", origaddr)
+		return
+	}
+	reg := r.source[origaddr]
+	reg.ClientName = cname
+	//fmt.Printf("SUBSCRIBE REGISTRY Client Name: [%s] Client ID: [%s] value : %+v \n", cname, origaddr, reg)
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/registry.go]: Remove

File: pkg/http-tunnel/server/registry.go [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+
+// Subscriber returns client identifier assigned to given host.
+func (r *registry) Subscriber(hostPort string) (string, *Auth, bool) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+	h, ok := r.hosts[trimPort(hostPort)]
+	if !ok {
+		return "", nil, false
+	}
+	fmt.Printf("SUBSCRIBER REGISTRY [%s] value : %+v \n", hostPort, h)
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/registry.go]: All of these commented and uncommented prints in this file should be logs or removed

File: pkg/http-tunnel/server/registry.go [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+func (r *registry) clear(identifier string) *RegistryItem {
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	ilogger := r.logger.WithFields(log.Fields{
+		"identifier": identifier,
+	})
+
+	//i, ok := r.items[identifier]
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/registry.go]: There is a lot of unnecessary comented out code that we should look into and erase

File: pkg/http-tunnel/server/server.go [hunk 13]
```diff
@@ -0,0 +1,762 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to create port range pool: %s", err)
+	}
+
+	listener, err := listener(serverData)
+	if err != nil {
+		return nil, fmt.Errorf("listener failed: %s", err)
+	}
+
+	logger := serverData.logger
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/server.go]: Same as client don't think we need theses as we did the same thing in NewClient
IMO in one place this logic is extra

File: pkg/http-tunnel/server/tcpproxy.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+		"src": target,
+	})
+	transfer(local, r, loggerWithContext)
+
+	<-done
+}
+
+func (p *TCPProxy) localAddrFor(hostPort string) string {
+
+	//	fmt.Printf("TCPPROXY localAddrFor GET FROM %s: %#+v\n ", hostPort, p.localAddrMap)
```
Comment (mpetrun5) [inline pkg/http-tunnel/server/tcpproxy.go]: Someone really liked commented out prints

**Review by mpetrunic — COMMENTED** [review]
- What's up with tls configs, can we remove those?
- Can you move tunnel logs from info to debug?

File: pkg/http-tunnel/client/client.go [hunk 15]
```diff
@@ -0,0 +1,334 @@
+// Copyright (C) 2017 Michał Matczuk
+// Use of this source code is governed by an AGPL-style
+// license that can be found in the LICENSE file.
+
+package client
+
+import (
+	"crypto/tls"
```
Comment (mpetrunic) [inline pkg/http-tunnel/client/client.go]: khm?

**Comment by MakMuftic** [comment]
@mpetrun5 @mpetrunic 
- removed tls connection
- moved all tunnel logs to debug
- added integration test (starts `client` and `server` and checks if the tunnel opened successfully)

### [NEW] Example
#### Creating server
```go 
s, err := server.NewServer(&server.ServerConfig{
	Address:        ":5223",
	PortRange:      "10000:50000",
	AuthHandler: func(s string) bool {
		return s == "test-token"
	},
	Logger: log.NewEntry(l),
})
if err != nil {
	l.Fatalf("failed to create server: %s", err)
}
s.Start()
```
#### Creating client
```go 
c, err := client.NewClient(&client.ClientConfig{
	ServerAddress:  "127.0.0.1:5223",
	Tunnels: map[string]*client.Tunnel{
		"": {
			Protocol:   "tcp",
			Addr:       "localhost:3000",
			Auth:       "",
			Host:       "",
			RemoteAddr: "0.0.0.0:AUTO",
		},
	},
	Logger:    log.NewEntry(l),
	AuthToken: "test-token",
})

if err != nil {
	l.Fatalf("failed to create client: %s", err)
}

if err := c.Start(); err != nil {
	l.Fatalf("failed to start tunnels: %s", err)
}
```

File: pkg/http-tunnel/client/utils.go [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+	return p
+}
+
+func CreateProxy(m map[string]*Tunnel, logger *logrus.Entry) ProxyFunc {
+	httpURL := make(map[string]*url.URL)
+	tcpAddr := make(map[string]string)
+
+	for v, t := range m {
+		fmt.Printf("Creating proxy for %#+v/%#+v\n", v, t)
```
Comment (mpetrunic) [inline pkg/http-tunnel/client/utils.go]: use logger?

File: pkg/http-tunnel/client/utils.go [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+		switch t.Protocol {
+		case proto.HTTP:
+			u, err := url.Parse(t.Addr)
+			if err != nil {
+				logger.Fatalf("invalid tunnel address: %v", err)
+			}
+			httpURL[t.Host] = u
+		case proto.TCP, proto.TCP4, proto.TCP6:
+			fmt.Printf("Setting config for %s | REMOTE: %s | LOCAL: %s\n", v, t.RemoteAddr, t.Addr)
```
Comment (mpetrunic) [inline pkg/http-tunnel/client/utils.go]: ditto

**Review by mpetrun5 — APPROVED** [review]

**Pull request merged**